### PR TITLE
Add cross entropy loss.

### PR DESCRIPTION
**Pull request by lorenzrichter**

### CODE REVIEW

**Review by zwpku — APPROVED** [review]
1. Thanks for including the cross-entropy functional. I will check it out.
2. On the variance, I do not really recommend to add it to the master. It may be useful to get a feeling about the optimization procedure, but it is also expensive and make the logic of the code less clear. Maybe it is better to use it in local branch for debug purpose.
